routers/api/actions/artifacts.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package actions
+
+import (
+	gocontext "context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"code.gitea.io/gitea/models/actions"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/storage"
+	"code.gitea.io/gitea/modules/web"
+)
+
+const artifactRouteBase = "/_apis/pipelines/workflows/{taskID}/artifacts"
+
+func ArtifactsRoutes(goctx gocontext.Context, prefix string) *web.Route {
+	m := web.NewRoute()
+	m.Use(withContexter(goctx))
+
+	r := artifactRoutes{
+		prefix: prefix,
+		fs:     storage.ActionsArtifacts}
+
+	// retrieve, list and confirm artifacts
+	m.Post(artifactRouteBase, r.getUploadArtifactURL)
+	m.Get(artifactRouteBase, r.listArtifacts)
+	m.Patch(artifactRouteBase, r.patchArtifact)
+
+	// handle container artifacts
+	m.Put(artifactRouteBase+"/{artifactID}/upload", r.uploadArtifact)
+	m.Get(artifactRouteBase+"/{artifactID}/path", r.getDownloadArtifactURL)
+	m.Get(artifactRouteBase+"/{artifactID}/download", r.downloadArtifact)
+
+	return m
+}
+
+// withContexter initializes a package context for a request.
+func withContexter(ctx gocontext.Context) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+			ctx := context.Context{
+				Resp: context.NewResponse(resp),
+				Data: map[string]interface{}{},
+			}
+			defer ctx.Close()
+
+			ctx.Req = context.WithContext(req, &ctx)
+
+			next.ServeHTTP(ctx.Resp, ctx.Req)
+		})
+	}
+}
+
+type artifactRoutes struct {
+	prefix string
+	fs     storage.ObjectStorage
+}
+
+func (ar artifactRoutes) openFile(fpath string, contentRange string) (storage.Object, bool, error) {
+	if contentRange != "" && !strings.HasPrefix(contentRange, "bytes 0-") {
+		f, err := ar.fs.Open(fpath)
+		if err != nil {
+			return nil, false, err
+		}
+		_, err = f.Seek(0, os.SEEK_END)
```

Comment:
Why it's `f.Seek(0, os.SEEK_END)`? When I read the code, I think the `Seek` should respect the `Content-Range`?

Reply:
When artifact uploading, it separates large file to several chunk with content-range as incremental uploading.
So we need seek to end and append current chunk at the end of file.

Reply:
I know, but I  think the Seek  should respect the `Content-Range`.  The uploader should tell the server the offset (position).

Reply:
Think about a case:  2  chunks, each  chunk is 100 bytes.

* Uploader  uploads chunk 1 , 100 bytes done
* Uploader  uploads chunk 2 , 20 bytes, then network error
* Uploader  retry to upload chunk 2 ,  then if you Seek to 120, the file is corrupted.

Reply:
I see. let me try to update this code. a little complicate.